dart/common/AddonManager.h

```diff
@@ -0,0 +1,176 @@
+/// constant-time access to specific Addon types, there are FOUR macros that you
+/// should use in your derived class:
+///
+/// DART_ENABLE_ADDON_SPECIALIZATION() must be declared once in the derived
+/// class's definition, under a 'public:' declaration range.
+///
+/// DART_SPECIALIZE_ADDON_INTERNAL( AddonType ) must be declared once for each
+/// AddonType that you want to specialize. It should be placed in the derived
+/// class's definition, under a 'public:' declaration range.
+///
+/// DART_SPECIALIZE_ADDON_EXTERNAL( Derived, AddonType ) must be declared once
+/// for each AddonType that you want to specialize. It should be placed
+/// immediately after the class's definition in the same header file as the
+/// derived class, inside of the same namespace as the derived class. This macro
+/// defines a series of templated functions, so it should go in a header, and
+/// not in a source file.
+///
+/// DART_INSTANTIATE_SPECIALIZED_ADDON( AddonType ) must be declared once for
+/// each AddonType that you want to specialize. It should be placed inside the
+/// constructor of the derived class, preferably before anything else is done
+/// inside the body of the constructor.
+class AddonManager
+{
+public:
+
+  using StateMap = std::map< std::type_index, std::unique_ptr<Addon::State> >;
+  using State = ExtensibleMapHolder<StateMap>;
+
+  using PropertiesMap = std::map< std::type_index, std::unique_ptr<Addon::Properties> >;
+  using Properties = ExtensibleMapHolder<PropertiesMap>;
+
+  using AddonMap = std::map< std::type_index, std::unique_ptr<Addon> >;
+
+  /// Virtual destructor
+  virtual ~AddonManager() = default;
+
+  /// Check if this AddonManager has a certain type of Addon
+  template <class T>
+  bool has() const;
+
+  /// Get a certain type of Addon from this AddonManager
+  template <class T>
+  T* get();
+
+  /// Get a certain type of Addon from this AddonManager
+  template <class T>
+  const T* get() const;
+
+  /// Make a clone of the addon and place the clone into this AddonManager. If
+  /// an Addon of the same type already exists in this AddonManager, the
+  /// existing Addon will be destroyed.
+  template <class T>
+  void set(const T* addon);
+
+  /// Use move semantics to place addon into this AddonManager. If an Addon of
+  /// the same type already exists in this AddonManager, the existing Addon will
+  /// be destroyed.
+  template <class T>
+  void set(std::unique_ptr<T>&& addon);
+
+  /// Construct an Addon inside of this AddonManager
+  template <class T, typename ...Args>
+  T* create(Args&&... args);
+
+  /// Remove an Addon from this AddonManager.
+  template <class T>
+  void erase();
+
+  /// Remove an Addon from this AddonManager, but return its unique_ptr instead
+  /// of letting it be deleted. This allows you to safely use move semantics to
+  /// transfer an Addon between two AddonManagers.
+  template <class T>
+  std::unique_ptr<T> release();
```

Comment:
I want to modify `set` to return the `unique_ptr` that was replaced. This is identical to the existing behavior if the user ignores the return value because it will be immediately destructed.